test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java

```diff
@@ -0,0 +1,242 @@
+    }
+
+    private void readEventsAndVerify(int startInclusive, int endExclusive) {
+        log.info("Read and Verify events between [{},{})", startInclusive, endExclusive);
+        List<EventRead<Integer>> eventsRead = readEvents();
+        verifyEvents(eventsRead, startInclusive, endExclusive);
+    }
+
+    private void verifyEvents(List<EventRead<Integer>> events, int startInclusive, int endExclusive) {
+
+        Supplier<Stream<Integer>> streamSupplier = () -> events.stream().map(i -> i.getEvent()).sorted();
+        IntSummaryStatistics stats = streamSupplier.get().collect(Collectors.summarizingInt(value -> value));
+
+        assertTrue("Check for first event", stats.getMin() == startInclusive);
+        assertTrue("Check for last event", stats.getMax() == endExclusive - 1);
+        //Check for missing events
+        assertEquals("Check for number of events", endExclusive - startInclusive, stats.getCount());
+        assertEquals("Check for duplicate events", endExclusive - startInclusive,
+                streamSupplier.get().distinct().count());
+    }
+
+    private <T extends Serializable> List<EventRead<T>> readEvents() {
+        List<EventRead<T>> events = new ArrayList<>();
+
+        try (ClientFactory clientFactory = ClientFactory.withScope(SCOPE, controllerURI);
+             EventStreamReader<T> reader = clientFactory.createReader("reader",
+                     READER_GROUP_NAME,
+                     new JavaSerializer<T>(),
+                     readerConfig)) {
+            log.info("Reading events from {}/{}", SCOPE, STREAM);
+            EventRead<T> event = null;
+            do {
+                try {
+                    event = reader.readNextEvent(READ_TIMEOUT);
+                    if (event.getEvent() != null) {
+                        log.info("Read event {}", event.getEvent());
+                        events.add(event);
+                    }
+                } catch (ReinitializationRequiredException e) {
+                    log.error("Exception while reading event", e);
+                    fail("Reinitialization Exception is not expected");
+                }
+            } while (event.getEvent() != null);
+            log.info("No more events from {}/{}", SCOPE, STREAM);
+        }
+        return events;
+    }
+
+    private <T extends Serializable> void writeEvents(List<T> events) {
+        try (ClientFactory clientFactory = ClientFactory.withScope(SCOPE, controllerURI);
+             EventStreamWriter<T> writer = clientFactory.createEventWriter(STREAM,
+                     new JavaSerializer<T>(),
+                     EventWriterConfig.builder().build())) {
+            for (T event : events) {
+                String routingKey = "constantKey"; // constant routing key
+                log.info("Writing message: {} with routing-key: {} to stream {}", event, routingKey, STREAM);
+                writer.writeEvent(routingKey, event);
+                writer.flush();
```

Comment:
This is not entirely related to the goal of this test case, but it is conservative to flush upon every write. If one of the writes fail, then all subsequent outstanding writes should also fail. I think we should either flush upon some given number of writes or simply flush at the end, which will happen automatically when we close the writer.

Reply:
done.